### PR TITLE
Train Dat Program To Read train.dat

### PR DESCRIPTION
**Pull request by cwfitzgerald**
I'm sorry.

This adds support for parsing train.dat using the standard kvp parser. Notable additions to be looked at:

- Added FromKVPValueEnumNumbers that takes an enum and creates it from a i64. It has the ability to change the index, but follows rust's enum numbering scheme and just increments from there.
- Made KVP parser generic over the symbols it looks for the start and end of a section, as well as the separator between key and value.
- Other than that the rest of the PR is a pretty bland transcription of the docs.

There will be hella bugs with this, but corpus testing is next.

### CODE REVIEW

File: bve/src/parse/train_dat/sections/device.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    #[kvp(bare)]
+    pass_alarm: PassAlarmType,
+    #[kvp(bare)]
+    door_open_mode: DoorMode,
+    #[kvp(bare)]
+    door_close_mode: DoorMode,
+}
+
+#[derive(Debug, Clone, PartialEq, FromKVPValueEnumNumbers)]
+pub enum AtsAvailable {
```
Comment (DethRaid) [inline bve/src/parse/train_dat/sections/device.rs]: Ats?

This abbreviation might be obvious to BVE people, but it's not obvious to me. Could you either expand the abbreviation in the enum name, or add a comment explaining it?

File: bve/src/parse/train_dat/sections/device.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    #[kvp(default, index = "-1")]
+    Neither,
+    AtsSn,
+    Both,
+}
+
+#[derive(Debug, Clone, PartialEq, FromKVPValueEnumNumbers)]
+pub enum AtcAvailable {
```
Comment (DethRaid) [inline bve/src/parse/train_dat/sections/device.rs]: Same as AtsAvailable

File: bve/src/parse/train_dat/sections/device.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    #[kvp(default)]
+    Unavailable,
+    Manual,
+    Automatic,
+}
+
+#[derive(Debug, Clone, PartialEq, FromKVPValueEnumNumbers)]
+pub enum EbAvailable {
```
Comment (DethRaid) [inline bve/src/parse/train_dat/sections/device.rs]: Same as AtsAvailable

File: bve/src/parse/train_dat/sections/device.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    Unavailable,
+    Available,
+}
+
+#[derive(Debug, Clone, PartialEq, FromKVPValueEnumNumbers)]
+pub enum ReAdhesionDeviceType {
+    #[kvp(default, index = "-1")]
+    Unavailable,
+    /// Cuts off power instantly and rebuilds it up fast in steps.
+    TypeA,
```
Comment (DethRaid) [inline bve/src/parse/train_dat/sections/device.rs]: Is this their actual names? Ugh engineers shouldn't be allowed to name things

File: bve/src/parse/train_dat/sections/motor.rs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+pub struct MotorSection {
+    #[kvp(bare)]
+    sound_data: Vec<MotorSoundType>,
+}
+
+#[derive(Debug, Clone, PartialEq, FromKVPValue)]
+pub struct MotorSoundType {
+    sound_index: i64,
+    /// Really just sound speed
+    pitch: f32,
```
Comment (DethRaid) [inline bve/src/parse/train_dat/sections/motor.rs]: If it's a sound speed and not a real pitch, does it have to be named `pitch`?

File: bve/src/parse/train_dat/sections/performance.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+use bve_derive::FromKVPSection;
+
+#[derive(Debug, Clone, PartialEq, FromKVPSection)]
+pub struct PerformanceSection {
+    #[kvp(bare)]
+    pub deceleration: f32,
+    #[kvp(bare)]
+    pub coefficient_of_static_friction: f32,
+    #[kvp(bare)]
+    pub _reserved0: f32,
```
Comment (DethRaid) [inline bve/src/parse/train_dat/sections/performance.rs]: unwu

File: bve/src/parse/train_dat/sections/version.rs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    pub version: Version,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Version {
+    BVE120,
+    BVE121,
+    BVE122,
+    BVE2,
+    OpenBVE { version: String },
```
Comment (DethRaid) [inline bve/src/parse/train_dat/sections/version.rs]: BVEReborn when

File: bve/src/parse/train_dat/sections/device.rs (hunk 1 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/device.rs]: No idea what it stands for. I'd have to go poking around. It's some kind of automatic signalling used in japan.

File: bve/src/parse/train_dat/sections/device.rs (hunk 2 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/device.rs]: ¯\_(ツ)_/¯

File: bve/src/parse/train_dat/sections/device.rs (hunk 3 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/device.rs]: Fine, that's the ebrake.

File: bve/src/parse/train_dat/sections/device.rs (hunk 4 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/device.rs]: Yerp

File: bve/src/parse/train_dat/sections/motor.rs (hunk 5 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/motor.rs]: That's what it's named in the docs. I'm trying to keep it fairly consistent with the docs.

File: bve/src/parse/train_dat/sections/performance.rs (hunk 6 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/performance.rs]: Backwards compatibility ¯\\\_(ツ)\_/¯

File: bve/src/parse/train_dat/sections/version.rs (hunk 7 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/version.rs]: Hopefully never as I don't want to use this damn file format lmao

File: bve-derive/src/kvp.rs [hunk 8]
```diff
@@ -299,3 +303,116 @@ pub fn kvp_section(item: TokenStream) -> TokenStream {
...
+                })
+            }
+        }
+    )
+    .into()
+}
+
+#[derive(Debug, FromVariant)]
+#[darling(attributes(kvp))]
+struct Variant {
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply bve-derive/src/kvp.rs]: This represents a variant in the provided enum. The FromVariant/darling stuff is about extracting data from the attributes. Note here that terminology is different from C++ to Rust. Rust calls the whole thing an enum with each possible option being a Variant, while c++ calls the whole thing a variant, and doesn't call the other thing anything as it's all template bullshit.

tl;dr: I'm not sure what other name to give it as it is neither bve specific, or an actual value.

File: bve/src/parse/train_dat/sections/device.rs (hunk 1 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/device.rs]: Added some docs

File: bve/src/parse/train_dat/sections/device.rs (hunk 2 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/device.rs]: As above

File: bve/src/parse/train_dat/sections/device.rs (hunk 3 above)
Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/device.rs]: Actually expanded the name

File: bve-derive/src/kvp.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply bve-derive/src/kvp.rs]: Added a mildly helpful comment

File: bve/src/parse/train_dat/sections/device.rs [hunk 9]
```diff
@@ -24,14 +24,16 @@ pub struct DeviceSection {
     door_close_mode: DoorMode,
 }
 
+/// Japanese implementation of digital Automatic Train Stops, supports ATS-SN and ATS-P
```
Comment (DethRaid) [inline bve/src/parse/train_dat/sections/device.rs]: Wait what about other countries' implementations?

Comment (cwfitzgerald) [reply bve/src/parse/train_dat/sections/device.rs]: Bring your own ats implementation as a dll plugin. Isn't bve great!
